File: Backend_Victorino/src/main/java/org/victorino_style/dto/NotificacionDto.java
package org.victorino_style.dto;

import org.victorino_style.entity.Notificacion;

import java.time.Instant;

// DTO de respuesta para la bandeja de notificaciones in-app.
// Evita exponer la entidad JPA directamente y sus relaciones LAZY.
public record NotificacionDto(

        Long id,

        // Titulo de la notificacion (max 250 caracteres).
        String titulo,

        // Cuerpo / mensaje de la notificacion (max 500 caracteres).
        String cuerpo,

        // Tipo: CONFIRMACION_RESERVA, CANCELACION_PELUQUERIA, RECORDATORIO_24H, etc.
        String tipo,

        // ID de la cita relacionada (puede ser null si no hay cita asociada).
        Long idCita,

        // true si Firebase confirmo la entrega del push.
        boolean enviadaPush,

        // Fecha en que se creo la notificacion.
        Instant fechaCreacion,

        // null = no leida. Con valor = ya fue leida.
        Instant fechaLectura

) {

    // -----------------------------------------------------------------------
    // Factory method: convierte la entidad JPA en este DTO.
    // Se usa en el Service para no exponer la entidad hacia afuera.
    // -----------------------------------------------------------------------
    public static NotificacionDto from(Notificacion n) {
        return new NotificacionDto(
                n.getId(),
                n.getTituloNotificacion(),
                n.getCuerpoNotificacion(),
                n.getTipoNotificacion(),
                n.getIdCitaRelacionadaNotificacion() != null
                        ? n.getIdCitaRelacionadaNotificacion().getId()
                        : null,
                n.getEnviadaPushNotificacion(),
                n.getFechaCreacionNotificacion(),
                n.getFechaLecturaNotificacion()
        );
    }
}

File: frontend_victorino/lib/features/notificaciones/data/modelos/notificacion_dto.dart
import '../../domain/entidades/notificacion.dart';

// DTO que mapea el JSON del backend al dominio de Flutter.
// El backend devuelve:
//   { "id", "titulo", "cuerpo", "tipo", "idCita", "enviadaPush",
//     "fechaCreacion", "fechaLectura" }
class NotificacionDto {
  const NotificacionDto({
    required this.id,
    required this.titulo,
    required this.cuerpo,
    required this.tipo,
    required this.enviadaPush,
    required this.fechaCreacion,
    this.idCita,
    this.fechaLectura,
  });

  final int id;
  final String titulo;
  final String cuerpo;
  final String tipo;
  final bool enviadaPush;
  final DateTime fechaCreacion;
  final int? idCita;
  final DateTime? fechaLectura;

  factory NotificacionDto.fromJson(Map<String, dynamic> json) {
    return NotificacionDto(
      id: json['id'] as int,
      titulo: json['titulo'] as String? ?? '',
      cuerpo: json['cuerpo'] as String? ?? '',
      tipo: json['tipo'] as String? ?? '',
      enviadaPush: json['enviadaPush'] as bool? ?? false,
      fechaCreacion: DateTime.parse(json['fechaCreacion'] as String),
      idCita: json['idCita'] as int?,
      fechaLectura: json['fechaLectura'] != null
          ? DateTime.parse(json['fechaLectura'] as String)
          : null,
    );
  }

  Notificacion aEntidad() {
    return Notificacion(
      id: id,
      titulo: titulo,
      cuerpo: cuerpo,
      tipo: tipo,
      enviadaPush: enviadaPush,
      fechaCreacion: fechaCreacion,
      idCita: idCita,
      fechaLectura: fechaLectura,
    );
  }
}

